crates/astria-composer/src/searcher.rs

```diff
@@ -0,0 +1,362 @@
+    // action belongs to.
+    rollup_chain_id: String,
+    // Channel to report the internal tatus of the searcher to other parts of the system.
+    status: watch::Sender<Status>,
+    // Set of currently running jobs converting pending eth transactions to signed sequencer
+    // transactions.
+    conversion_tasks: JoinSet<eyre::Result<SignedSequencerTx>>,
+    // Set of in-flight RPCs submitting signed transactions to the sequencer.
+    // submission_tasks: JoinSet<eyre::Result<tx_sync::Response>>,
+    submission_tasks: JoinSet<eyre::Result<()>>,
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct Status {
+    eth_connected: bool,
+    sequencer_connected: bool,
+}
+
+impl Status {
+    pub(crate) fn is_ready(&self) -> bool {
+        self.eth_connected && self.sequencer_connected
+    }
+}
+
+/// A thin wrapper around [`Provider<Ws>`] to add timeouts.
+///
+/// Currently only provides a timeout around for `get_net_version`.
+/// TODO(https://github.com/astriaorg/astria/issues/216): add timeouts for
+/// `subscribe_full_pendings_txs` (more complex because it's a stream).
+#[derive(Clone)]
+struct EthClient {
+    inner: Provider<Ws>,
+}
+
+impl EthClient {
+    async fn connect(url: &str) -> Result<Self, ProviderError> {
+        let inner = Provider::connect(url).await?;
+        Ok(Self {
+            inner,
+        })
+    }
+
+    /// Wrapper around [`Provider::get_net_version`] with a 1s timeout.
+    async fn get_net_version(&self) -> eyre::Result<String> {
+        use ethers::providers::Middleware as _;
+        tokio::time::timeout(Duration::from_secs(1), self.inner.get_net_version())
+            .await
+            .wrap_err("request timed out")?
+            .wrap_err("RPC returned with error")
+    }
+}
+
+/// A thin wrapper around [`sequencer_client::Client`] to add timeouts.
+///
+/// Currently only provides a timeout for `abci_info`.
+#[derive(Clone)]
+struct SequencerClient {
+    inner: sequencer_client::HttpClient,
+}
+
+impl SequencerClient {
+    #[instrument]
+    fn new(url: &str) -> eyre::Result<Self> {
+        let inner = sequencer_client::HttpClient::new(url)
+            .wrap_err("failed to construct sequencer client")?;
+        Ok(Self {
+            inner,
+        })
+    }
+
+    /// Wrapper around [`Provider::get_net_version`] with a 1s timeout.
+    async fn abci_info(self) -> eyre::Result<abci::response::Info> {
+        use sequencer_client::Client as _;
+        tokio::time::timeout(Duration::from_secs(1), self.inner.abci_info())
+            .await
+            .wrap_err("request timed out")?
+            .wrap_err("RPC returned with error")
+    }
+}
+
+impl Searcher {
+    /// Constructs a new Searcher service from config.
+    ///
+    /// # Errors
+    ///
+    /// Errors are returned in the following scenarios:
+    /// + failed to connect to the eth RPC server;
+    /// + failed to construct a sequencer clinet
+    pub(super) async fn from_config(cfg: &Config) -> eyre::Result<Self> {
```

Comment:
curious why `pub(super)` vs `pub(crate)` here?

Reply:
It's more restricted than making it available crate-wide. No reason other than that honestly.